### ISSUE TITLE
"Unable to load .NET runtime, specifically hostfxr" due to Godot not seeing manually installed .NET SDK

### ISSUE THREAD

**Issue by carlos-montes-karasu** [godotengine/godot#75668]
### Godot version

4.0.2

### System information

macOS Ventura 13.3 - M1 CPU

### Issue description

I just installed .Net SDK (tried with 7.0 and 6.0 and same result)
After download Godot .NET I opened it and then an error dialog appear the title

"Failed to Load .Net runtime" and the body
"Unable to load .NET runtime, specifically hostfxr"

I do not know what else I can do, I can access `dotnet` command from my terminal, so I guess it was installed right

### Steps to reproduce

- just download godot .net for mac
- try to open it

### Minimal reproduction project

N/A

**Comment by Calinou** [comment]
- Duplicate of https://github.com/godotengine/godot/issues/75186.

**Comment by tkepus** [comment]
I have the exact same problem, except on the PC.  Installed the latest SDK that matches my architecture (x64) and Godot still gives me the 'missing runtime' error.  

I'm able to run .net applications, run dotnet, create a .net application, etc., but godot still fails to load the .net runtime...

**Comment by faijeya** [comment]
Reproduced and maybe solved on Steam Deck.
tl;dr: `export PATH=$PATH:~/.dotnet && export DOTNET_ROOT=~/.dotnet && ./Godot_v4.0.2-stable_mono_linux.x86_64`


Steam Deck by default offers you an immutable system with everything of note installed via flatpaks or left in `~/`
Official Microsoft .NET installer indeed left all it installed at `~/.dotnet`
Being immutable, whatever env variables did the .NET installer set up is lost as tears in rain.

Now, _"Unable to load .NET runtime, specifically hostfxr"_ is fixed by setting `DOTNET_ROOT` Godot explicitly looks for (https://github.com/godotengine/godot/blob/3683b040eda27f33fece7a34fd1332572625c672/modules/mono/editor/hostfxr_resolver.cpp#L298)
For me it's `export DOTNET_ROOT=~/.dotnet`

Then the editor will start, but will emit helpful errors like _"ERROR: .NET Sdk not found. The required version is '6.0.15'.  at: godotsharp_pusherror (modules/mono/glue/runtime_interop.cpp:1303)"_
After a bit of playing around it became clear that the source of it is in FindDotNetExe() (https://github.com/godotengine/godot/blob/0a0843a67d51eaacf5d3350d6c39a3893e3c1d6c/modules/mono/editor/GodotTools/GodotTools/Build/DotNetFinder.cs#L15)

For MacOS it seems like it looks in a couple of hardcoded locations, so nothing you can do except maybe symlink `/usr/local/share/dotnet/` to wherever your dotnet executable is located.
Dotnet executable should likely be named `dotnet` and respond with non-gibberish when launched as `./dotnet --list-sdks`
For other OS it looks for `dotnet` in PATH: takes every directory in PATH and looks in it.
So I just do `export PATH=$PATH:~/.dotnet`


Now I'm unsure whether it solves-solves the problem as I just started setting up Godot and all sample mono projects are for 3.x, upgrade to 4.x without question and then fail to run in a sea of red ink, but that's likely normal.

**Comment by super-continent** [comment]
Having this same issue on Windows 10 x64, tried overriding the environment variables and stuff, didn't fix it

**Comment by richtaur** [comment]
I was getting this error too (OS X 12.3). Here's what worked for me:

1. Go here: https://dotnet.microsoft.com/en-us/download/dotnet/6.0 
2. Download SDK 6.0.408 (for me it was macOS / Arm64)
3. Install
4. Restart
5. Godot no longer complains when it opens

**Comment by shinspiegel** [comment]
> Reproduced and maybe solved on Steam Deck. tl;dr: `export PATH=$PATH:~/.dotnet && export DOTNET_ROOT=~/.dotnet && ./Godot_v4.0.2-stable_mono_linux.x86_64`

> Now, _"Unable to load .NET runtime, specifically hostfxr"_ is fixed by setting `DOTNET_ROOT` Godot explicitly looks for (
> https://github.com/godotengine/godot/blob/3683b040eda27f33fece7a34fd1332572625c672/modules/mono/editor/hostfxr_resolver.cpp#L298
> 
> )

My `DOTNET_ROOT` looks correct, but still fails.
Could you kindly elaborate more on the issue with the macOS?
On the macOS it does not make the lookup for this env variable @faijeya?

**Comment by shinspiegel** [comment]
I'm not sure if this is correct, but I added a new entry for the lookout on the DotNotFinder class. Maybe this could be the solution to this issue. 

This is the PR,
https://github.com/godotengine/godot/pull/78994

**Comment by shinspiegel** [comment]
TLDR: `ln -s /opt/homebrew/opt/dotnet/libexec /usr/local/share/dotnet`

---

For now what I've done to solve the issue on macOS with the dotnet installed from the `brew`,

Need to create the symbolic link for the `libexec` path, and not for the `dotnet` itself. In this case I had to use the already created symlink from the `brew` located on `/opt/homebrew/opt/dotnet`. 

But it requres the `libexec` folder, so small adjustment. The final symlink can be seen on the TLDR.

**Comment by faijeya** [comment]
@shinspiegel 
> elaborate more on the issue with the macOS?
Sure.

I'm completely unfamiliar with MacOS, but the code for MacOS there in DotNetFinder.cs checks for:
* /usr/local/share/dotnet/x64/dotnet existence if the arch is x64
* /usr/local/share/dotnet/dotnet existence if the arch is not x64 (Apple Silicon likely)
That's pretty much it, a single valid location per arch.

I'd propose to just collapse [FindDotNetExe()](https://github.com/godotengine/godot/blob/0a0843a67d51eaacf5d3350d6c39a3893e3c1d6c/modules/mono/editor/GodotTools/GodotTools/Build/DotNetFinder.cs#L15) to be just `return OS.PathWhich("dotnet");` and add a MacOS-specific derivative of [OS.WhichPath()](https://github.com/godotengine/godot/blob/b7c2fd2e9a7f01644e15fef86083a3e0e0221e4d/modules/mono/editor/GodotTools/GodotTools/Utils/OS.cs#L168).
Least surprise and all that.

**Comment by raulsntos** [comment]
I'm also not very familiar with MacOS, but as far as I know the issue is that MacOS bundles aren't allowed to read user's environment variables which means we won't be able to read `PATH` or `DOTNET_ROOT`.

This means MacOS will only be able to find .NET if it's installed in the standard location (i.e.: using the installer from https://get.dot.net). It seems that homebrew doesn't install .NET to the standard location[^1], so creating a symbolic link would be the only way to solve this.

We also look for a file in `/etc/dotnet/install_location` (and `/etc/dotnet/install_location_x64` for 64-bit OS), but I'm not sure if MacOS allows us to read this file. If it does, that would be another way to solve it. The file should contain a single line with the path to the .NET root directory (same as the `DOTNET_ROOT` environment variable).

By the way, how to search for .NET is documented in https://github.com/dotnet/designs/blob/main/accepted/2020/install-locations.md#host-behavior-for-30. We also look for .NET in `PATH`, but that's just added as a fallback in case the standard method fails.

[^1]: There's a PR that adds the hombrew install location to the search: https://github.com/godotengine/godot/pull/74221.

**Comment by shinspiegel** [comment]
> This means MacOS will only be able to find .NET if it's installed in the standard location (i.e.: using the installer from https://get.dot.net). It seems that homebrew doesn't install .NET to the standard location[1](#user-content-fn-1-cf659af00fba0a65afe7125425584f0b), so creating a symbolic link would be the only way to solve this.

The issue with this is,   
The Microsoft installer is a hard installer, they access and modify things outside the `userland`, 
The same can be done with `brew`, which does not make things outside `userland`. 

Maybe we could add a footnote on the docs to warn `macOS` users regarding the usage of brew and creating the symlink.

What do you folks thing about it?

**Comment by JohnDorsey** [comment]
> Reproduced and maybe solved on Steam Deck. tl;dr: `export PATH=$PATH:~/.dotnet && export DOTNET_ROOT=~/.dotnet && ./Godot_v4.0.2-stable_mono_linux.x86_64`

I'm trying to launch Godot 4.1.1 mono on ubuntu 20.04. This gets me closer to success, but it still fails to load. This same installation of dotnet does not cause any errors when Godot 3.5.1 mono is launched.

$ export PATH=$PATH:~/.dotnet && export DOTNET_ROOT=~/.dotnet && ~/Desktop/Godot_v4.1.1-stable_mono_linux.x86_64
Godot Engine v4.1.1.stable.mono.official.bd6af8e0e - https://godotengine.org
/lib/x86_64-linux-gnu/libxkbcommon.so.0: undefined symbol: xkb_utf32_to_keysym
/lib/x86_64-linux-gnu/libxkbcommon.so.0: undefined symbol: xkb_keymap_key_get_mods_for_level
OpenGL API 3.3.0 NVIDIA 470.161.03 - Compatibility - Using Device: NVIDIA - NVIDIA GeForce GTX 1060 with Max-Q Design
 
Failed to load /home/john/.dotnet/shared/Microsoft.NETCore.App/6.0.11/libhostpolicy.so, error: /home/john/.dotnet/shared/Microsoft.NETCore.App/6.0.11/libhostpolicy.so: file too short
An error occurred while loading required library libhostpolicy.so from [/home/john/.dotnet/shared/Microsoft.NETCore.App/6.0.11]
Failed to initialize context for config: /home/john/BinInstalls/game_engines/godot/Godot_v4.1.1-stable_mono_linux_x86_64/GodotSharp/Api/Debug/GodotPlugins.runtimeconfig.json. Error code: 0x80008082
ERROR: hostfxr_initialize_for_runtime_config failed with code: -2147450750
   at: initialize_hostfxr_for_config (modules/mono/mono_gd/gd_mono.cpp:200)

**Comment by UltimaTerra** [comment]
I have a similar issue, same error as above, but I do have .NET via snap. Here is my environment variable.
export DOTNET_ROOT=/snap/dotnet-sdk/228/sdk/8.0.100

I just checked my profile rc, and this issue is kind of weird. I just caved and am going to work with vanilla godot with gdscript and hope I can add gdextension with a language I am more comfortable (I don't need C# much, just wanted to report this issue as well.) This is on Pop OS.

**Comment by jefetienne** [comment]
> We also look for a file in `/etc/dotnet/install_location` (and `/etc/dotnet/install_location_x64` for 64-bit OS),

I managed to fix running Godot 4.2.1 .NET edition on my Steam Deck (after a [manual binary download](https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/sdk-6.0.419-linux-x64-binaries) of .NET 6 to $HOME/.dotnet) by running these two commands:

```bash
sudo mkdir -p /etc/dotnet
sudo ln -sT ~/.dotnet /etc/dotnet/install_location
```

My .bashrc contains these two lines:

```bash
export DOTNET_ROOT="$HOME/.dotnet"
export PATH=$PATH:$DOTNET_ROOT:$DOTNET_ROOT/tools
```

**Comment by alexkotusenko** [comment]
I had the same issue on Mac M1. I downloaded .NET from the Microsoft website and set it up in my $PATH. It didn't work and it turned out that I downloaded the wrong macOS version (Intel instead of Silicon). I was able to open Godot with no errors once I downloaded the correct one and set it up in my PATH.

I hope this helps someone

**Comment by tomCarnevale** [comment]
I am having this issue with a fresh download from the godot website. 

Godot 4 c# support
MacOS Sonoma 14.1
Apple Silicon

I have my dotnet executable in the `/usr/local/share/dotnet/` folder, named dotnet.

running `dotnet --info` yields the following.

```
.NET SDK (reflecting any global.json):
 Version:   6.0.425
 Commit:    fcce060d9f

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  14.1
 OS Platform: Darwin
 RID:         osx-arm64
 Base Path:   /usr/local/share/dotnet/sdk/6.0.425/

global.json file:
  Not found

Host:
  Version:      6.0.33
  Architecture: arm64
  Commit:       6c636980f7

.NET SDKs installed:
  2.1.805 [/usr/local/share/dotnet/sdk]
  3.1.100 [/usr/local/share/dotnet/sdk]
  3.1.102 [/usr/local/share/dotnet/sdk]
  3.1.200 [/usr/local/share/dotnet/sdk]
  3.1.201 [/usr/local/share/dotnet/sdk]
  3.1.401 [/usr/local/share/dotnet/sdk]
  3.1.402 [/usr/local/share/dotnet/sdk]
  3.1.404 [/usr/local/share/dotnet/sdk]
  6.0.300 [/usr/local/share/dotnet/sdk]
  6.0.425 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.17 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.17 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.7 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.8 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.10 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.33 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.13 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.14 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.15 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.17 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.21 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.22 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.7 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.8 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.10 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.33 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.13 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
```

I've restarted, as suggested, to no avail. 

I've checked brew, it has never been installed that way. `brew uninstall dotnet` reminds me that its not installed. So i'm not entirely sure what is failing here.

**Comment by y0n1** [comment]
TL;DR: I solved this issue by creating a symlink.

```
sudo ln -s $DOTNET_ROOT /usr/local/share/dotnet 
```

In v4.3 the [hostfxr_resolver.cpp](https://github.com/godotengine/godot/blob/33c02a5c6bc2a050aad0e7356ebfcdb2592f54e0/modules/mono/editor/hostfxr_resolver.cpp#L201) module assumes, in macos, that the dotnet root directory is located at `/usr/local/share/dotnet`, this is the path where it gets installed when you install dotnet from the official web site. But the Hombrew installation instructs users to set the DOTNET_ROOT environment variable to `/opt/hombrew/opt/dotnet`.
IMHO this shouldn't be hardcoded but rather discovered from the DOTNET_ROOT environment variable or by executing `dotnet --info` and parsing the output.
